### PR TITLE
[SPARK-20798] GenerateUnsafeProjection should check if a value is null before calling the getter

### PR DESCRIPTION
**Pull request by ala**
## What changes were proposed in this pull request?

GenerateUnsafeProjection.writeStructToBuffer() did not honor the assumption that the caller must make sure that a value is not null before using the getter. This could lead to various errors. This change fixes that behavior.

Example of code generated before:
```scala
/* 059 */         final UTF8String fieldName = value.getUTF8String(0);
/* 060 */         if (value.isNullAt(0)) {
/* 061 */           rowWriter1.setNullAt(0);
/* 062 */         } else {
/* 063 */           rowWriter1.write(0, fieldName);
/* 064 */         }
```

Example of code generated now:
```scala
/* 060 */         boolean isNull1 = value.isNullAt(0);
/* 061 */         UTF8String value1 = isNull1 ? null : value.getUTF8String(0);
/* 062 */         if (isNull1) {
/* 063 */           rowWriter1.setNullAt(0);
/* 064 */         } else {
/* 065 */           rowWriter1.write(0, value1);
/* 066 */         }
```

## How was this patch tested?

Adds GenerateUnsafeProjectionSuite.

### CODE REVIEW

**Comment by ala** [comment]
@hvanhovell

**Comment by hvanhovell** [comment]
LGTM - pending jenkins

**Comment by hvanhovell** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala [hunk 1]
```diff
@@ -50,10 +50,15 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
...
-      val isNull = s"$input.isNullAt($i)"
-      ExprCode(code, isNull, fieldName)
+      val javaType = ctx.javaType(dt)
+      val isNullVar = ctx.freshName("isNull")
+      val valueVar = ctx.freshName("value")
+      val defaultValue = ctx.defaultValue(dt)
+      val readValue = ctx.getValue(input, dt, i.toString)
+      val code = s"""
+               boolean $isNullVar = $input.isNullAt($i);
+               $javaType $valueVar = $isNullVar ? $defaultValue : $readValue;"""
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: Nit: style issue. Could you fix the indentation? Thanks!

Comment (ala) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: Fixed. Now it looks like the rest of the file.

**Comment by hvanhovell** [comment]
LGTM - merging to master. Thanks!

**Pull request closed without merging**